client/src/components/Form/FormCardTool.vue

```diff
@@ -0,0 +1,280 @@
+            return this.options.versions;
+        },
+        showAddFavorite() {
+            return !!this.user.email && !this.isFavorite;
+        },
+        showVersions() {
+            return !this.sustainVersion && this.versions && this.versions.length > 1;
+        },
+        availableVersions() {
+            const versions = this.versions.slice();
+            const index = versions.indexOf(this.version);
+            if (index > -1) {
+                versions.splice(index, 1);
+            }
+            return versions.reverse();
+        },
+        showDownload() {
+            return this.user && this.user.is_admin;
+        },
+        showLink() {
+            return this.options.sharable_url;
+        },
+        isFavorite() {
+            return this.getFavorites().tools.includes(this.id);
+        },
+    },
+    created() {
+        // add tool menu webhooks
+        Webhooks.load({
+            type: "tool-menu",
+            callback: function (webhooks) {
+                webhooks.each((model) => {
+                    const webhook = model.toJSON();
+                    if (webhook.activate && webhook.config.function) {
+                        this.webhookDetails.push({
+                            icon: `fa ${webhook.config.icon}`,
+                            title: webhook.config.title,
+                            onclick: function () {
+                                const func = new Function("options", webhook.config.function);
+                                func(this.options);
+                            },
+                        });
+                    }
+                });
+            },
+        });
+    },
+    methods: {
+        onAddFavorite() {
+            axios.put(`${getAppRoot()}api/users/${this.user.id}/favorites/tools`, { object_id: this.id }).then(
+                (response) => {
+                    this.errorText = null;
+                    this.updateFavorites("tools", response.data);
+                    ariaAlert("added to favorites");
+                },
+                () => {
+                    this.errorText = `Failed to add '${this.id}' to favorites.`;
+                    ariaAlert("failed to add to favorites");
+                }
+            );
+        },
+        onRemoveFavorite() {
+            axios
+                .delete(`${getAppRoot()}api/users/${this.user.id}/favorites/tools/${encodeURIComponent(this.id)}`)
+                .then(
+                    (response) => {
+                        this.errorText = null;
+                        this.updateFavorites("tools", response.data);
+                        ariaAlert("removed from favorites");
+                    },
+                    () => {
+                        this.errorText = `Failed to remove '${this.id}' from favorites.`;
+                        ariaAlert("failed to remove from favorites");
+                    }
+                );
+        },
+        onCopyLink() {
+            copy(
```

Comment:
Unit testing will be possible if you externalize operations that use "window", "location", because those objects do not exist in the node unit test environment. Just put this behaviors that use window or other browser-global objects in a separate function, then import it. If you do that, when you write your unit tests you can just mock the imported function.

Reply:
This was unchanged from the previous code, moved it to a separate utility within this PR.